core/lib/Cardano/Address/Style/Shared.hs

```diff
@@ -0,0 +1,323 @@
+-- $overview
+--
+-- This module provides an implementation of:
+--
+-- - 'Cardano.Address.Derivation.GenMasterKey': for generating Shared master keys from mnemonic sentences
+-- - 'Cardano.Address.Derivation.HardDerivation': for hierarchical hard derivation of parent to child keys
+-- - 'Cardano.Address.Derivation.SoftDerivation': for hierarchical soft derivation of parent to child keys
+--
+-- - 'paymentAddress': for constructing payment addresses from a address public key or a script
+-- - 'delegationAddress': for constructing delegation addresses from payment credential (public key or script) and stake credential (public key or script)
+-- - 'pointerAddress': for constructing delegation addresses from payment credential (public key or script) and chain pointer
+-- - 'stakeAddress': for constructing reward accounts from stake credential (public key or script)
+
+-- | A cryptographic key for sequential-scheme address derivation, with
+-- phantom-types to disambiguate key types.
+--
+-- @
+-- let rootPrivateKey = Shared 'RootK XPrv
+-- let accountPubKey  = Shared 'AccountK XPub
+-- let addressPubKey  = Shared 'PaymentK XPub
+-- @
+--
+-- @since 3.3.0
+newtype Shared (depth :: Depth) key = Shared
+    { getKey :: key
+        -- ^ Extract the raw 'XPrv' or 'XPub' wrapped by this type.
+        --
+        -- @since 3.4.0
+    }
+    deriving stock (Generic, Show, Eq)
+
+deriving instance (Functor (Shared depth))
+instance (NFData key) => NFData (Shared depth key)
+
+--
+-- Key Derivation
+--
+-- $keyDerivation
+--
+-- === Generating a root key from 'SomeMnemonic'
+-- > :set -XOverloadedStrings
+-- > :set -XTypeApplications
+-- > :set -XDataKinds
+-- > import Cardano.Mnemonic ( mkSomeMnemonic )
+-- >
+-- > let (Right mw) = mkSomeMnemonic @'[15] ["network","empty","cause","mean","expire","private","finger","accident","session","problem","absurd","banner","stage","void","what"]
+-- > let sndFactor = mempty -- Or alternatively, a second factor mnemonic transformed to bytes via someMnemonicToBytes
+-- > let rootK = genMasterKeyFromMnemonic mw sndFactor :: Shared 'RootK XPrv
+--
+-- === Deriving child keys
+--
+-- Let's consider the following 3rd, 4th and 5th derivation paths @0'\/0\/14@
+--
+-- > let Just accIx = indexFromWord32 0x80000000
+-- > let acctK = deriveAccountPrivateKey rootK accIx
+-- >
+-- > let Just addIx = indexFromWord32 0x00000014
+-- > let addrK = deriveAddressPrivateKey acctK UTxOExternal addIx
+--
+-- > let stakeK = deriveDelegationPrivateKey acctK
+
+instance Internal.GenMasterKey Shared where
+    type SecondFactor Shared = ScrubbedBytes
+
+    genMasterKeyFromXPrv = liftXPrv
+    genMasterKeyFromMnemonic fstFactor sndFactor =
+        Shared $ generateNew seedValidated sndFactor
+        where
+            seed  = someMnemonicToBytes fstFactor
+            seedValidated = assert
+                (BA.length seed >= minSeedLengthBytes && BA.length seed <= 255)
+                seed
+
+instance Internal.HardDerivation Shared where
```

Comment:
This would make sense indeed, especially because the rationale behind CIP-1854' really is to stick to the same structure ass CIP-1852'. So if we add some new role to CIP-1852' (e.g. Catalyst Voting Key) we want to also have it in CIP-1854'. Therefore, we do want to share the underlying code as much as possible.

Reply:
reused Shelley and Shared code when it comes to generation and derivation